### PR TITLE
added more unit tests

### PR DESCRIPTION
**Pull request by alexthe2**
## Changes

_Please list issues fixed by this PR here, using format "Fixes #the-issue-number"._

## Checklist

- [ ] An issue to be fixed by this PR is listed above.
- [ ] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [ ] The code generator is run and the code is formatted (via `just precommit`).
- [ ] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [ ] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by alexthe2** [comment]
I got some issues in one of my projects, that I can’t reproduce in the bridge, but maybe one of these tests will fail

**Comment by alexthe2** [comment]
If they don’t fail I just added more tests, but I’ll try a bit more, so don’t merge it yet

**Comment by alexthe2** [comment]
found the broken part, mirrors. They seem to break everything

**Comment by alexthe2** [comment]
If you don’t mind @fzyzcjy , can you explain me quickly how mirrors work exactly / is there some docs about them

**Comment by fzyzcjy** [comment]
https://cjycode.com/flutter_rust_bridge/feature/lang_external.html

Also cc @Unoqwy who implemented the good mirror feature

**Comment by alexthe2** [comment]
Finally got the tests to reproduce exactly my issue, will try to fix it now

**Comment by alexthe2** [comment]
So the problem is that when an Enum is mirrored, its data types are stayed as regular, not the mirrored structs

**Comment by alexthe2** [comment]
I got a first draft that seems to work, but still have 2 edge cases I need to figure out, list of enums and enums in pai.rs

**Comment by alexthe2** [comment]
Ok so in the current master version list of mirrored enums don’t work either, meaning that my changes right now have at least fixed enums that contain mirrored structs, I still have to do some refactoring but overall I’m finished with this part, list of enums will be a seperate pull request, let’s approach this one at a time

**Comment by fzyzcjy** [comment]
Looks great! I will probably review today :)

**Comment by alexthe2** [comment]
> Looks great! I will probably review today :)

I need to sleep 😴 , and some of the code requires a bit of rethinking, so when u review ignore this function:
```rust
fn extract_type(fields: &Fields) -> Option<String> {
        let f = match fields {
            Fields::Named(_) => None,
            Fields::Unnamed(u) => Some(u.to_token_stream()),
            Fields::Unit => None,
        }
        .map(|s| {
            s.to_string()
                .trim_start_matches('(')
                .trim_end_matches(')')
                .to_string()
        });

        if f.is_some()
            && vec!["String", "bool", "i32", "i64", "f32", "f64"]
                .contains(&f.as_ref().unwrap().as_str())
        {
            return None;
        }

        f
    }
```

I need to reorder it

**Comment by alexthe2** [comment]
Nvm, I guess. It was a very short refactor. If you can review it, it would be great. I can then change it later today (after I finally sleep).

If there are no nits and all tests pass(which I don't believe), then for the changelog would be:
- Bumped up syn, to latest v1 version
- Fixed enum containing structs
- Added a bunch of unit tests to make sure it works for the future (because it worked before, so some feature/refactor must have broken it)

**Comment by fzyzcjy** [comment]
Sure, take your time! Btw for the changelog - could you please modify the PR title, such that it is easier to read for future readers :)

File: frb_example/pure_dart/dart/lib/main.dart [hunk 1]
```diff
@@ -628,6 +628,44 @@ void main(List<String> args) async {
   //   expect(output.type, "not a type ;')");
   // });
 
+  test('test dart test more than jus ton eraw string struct', () async {
```
Comment (fzyzcjy) [inline frb_example/pure_dart/dart/lib/main.dart:631]:
```suggestion
  test('test dart test more than just one raw string struct', () async {
```

File: frb_codegen/src/ir/field.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ crate::ir! {
 pub struct IrField {
     pub ty: IrType,
     pub name: IrIdent,
+    pub wrapper_type: Option<String>,
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:10]: nit: May I know what is "wrapper_type" vs "ty", or can we infer the wrapper_type from the ty?

If we need to keep this field, would be great to add a bit of comments, explaining to readers this difference

**Review by fzyzcjy — COMMENTED** [review]
Nice fix! Not checked into detailed logic yet, want to know a bit about "wrapper_type" first

File: frb_codegen/src/ir/field.rs (hunk 2 above)
Comment (alexthe2) [reply frb_codegen/src/ir/field.rs:10]: depends, It could maybe be a simple bool, I can give it a try

**Comment by alexthe2** [comment]
how does that look?, not happy with the hardcoding though

**Comment by fzyzcjy** [comment]
Quite busy now, I will check it soon (probably within a few hours) :)

**Comment by alexthe2** [comment]
> Quite busy now, I will check it soon (probably within a few hours) :)

don’t worry, not pressing (maybe a bit because everything is broken on my side in the project)

**Comment by fzyzcjy** [comment]
> maybe a bit because everything is broken on my side in the project

Ah then I should do it quicker!

File: frb_codegen/src/generator/rust/ty_enum.rs [hunk 3]
```diff
@@ -232,7 +232,16 @@ impl TypeRustGeneratorTrait for TypeEnumRefGenerator<'_> {
                                     self.context.ir_file,
                                     self.context.config,
                                 );
-                                gen.convert_to_dart(field.name.rust_style().to_owned())
+                                let struct_name = field.ty.mirrored_nested();
+                                if field.mirrored_enum && struct_name.is_some() {
+                                    format!(
+                                        "mirror_{}({}).into_dart()",
+                                        struct_name.unwrap(),
+                                        field.name.rust_style().to_owned()
```
Comment (fzyzcjy) [inline frb_codegen/src/generator/rust/ty_enum.rs:240]: Disclaimer: Busy with my code so the suggestions may be inperfect!

Hmm if I understand correctly, this fix is like: The `IrField.ty`'s type name is the original struct name, instead of the mirrored struct name, but we want the latter. Thus, here we explicitly hack it to add `mirror_` prefix.

However, this looks like a general problem: What if other places also wants the mirrored name instead of the old name? But the current solution adds `mirrored_enum` flag *inside* IrField.

Ideally, maybe we can have somewhere *general* (instead of tied to IrField) showing whether a type is a mirrored enum. For example, can we ad a field in `IrType`?

Then we can do sth like

```
impl IrType {
  fn the_name_with_possibly_mirror() { ... }
}
```

and use it here happily.

Or, if we are sure every place wants the mirror instead of original type, then just make `the_name_with_possibly_mirror` the default choice, e.g. name it as `name()`

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs:240]: Hmm, ok I’ll look into that

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs:240]: The issue is solely for enums, so while I get the idea, I don’t think it’s useful to change the entire structure. I added a bunch more tests to make sure

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: what do you think of this way? I added the settings field, cause it feels a bit more *clean* then starting to bloat the IrField

Comment (fzyzcjy) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Hmm I am still a bit confused.

The current code is: For each *field*, if the *type* of the field is indeed a mirror, do extra log to code generation (addd "mirror_").

However, we know it is *type* that can be a mirror. So what I proposes is: Is it possible we add `is_mirror` or something like that onto the *type*, instead of adding it as a flag for the *field*?

**Comment by alexthe2** [comment]
Ohh, I get what you mean. I can change that very quickly

File: frb_codegen/src/generator/rust/ty_enum.rs (hunk 3 above)
Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Nvm, this is impossible because type is an enum ;)

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Technically it is possible, but that would be a lot of code, for an edge case. Because then it would have to become a type trait…, and be implemented for all of them. and then again this is only an issue for enums but the code generation would change a lot

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: should I change it, or are we sticking to this approach?

Comment (fzyzcjy) [reply frb_codegen/src/generator/rust/ty_enum.rs]: > that would be a lot of code, for an edge case

IMHO it should be only a few dozen lines of code.

Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Yes and no, the trait implementation would have to be for all types right? + The current ty approach would have to change to be changed at a later point (that's rather fine)

Comment (fzyzcjy) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Let me use a few minutes and see whether I can make a patch demonstrating my idea

**Comment by fzyzcjy** [comment]
Goal: If a type is a mirror, the name of it should be `mirror_ThatType` instead of `ThatType`.

Code prototype:

[TODO]

File: frb_codegen/src/generator/rust/ty_enum.rs (hunk 3 above)
Comment (alexthe2) [reply frb_codegen/src/generator/rust/ty_enum.rs]: Ofc, I just took another look myself but it would mean that:
1. New function in IrType, that essentially does the same mirrored_nested (easily doable)
 a. Added field in StructRef, that is then used by the function
2. The current logic does more or less sth in this way:
```rust
let gen = TypeRustGenerator::new(
    field.ty.clone(),
    self.context.ir_file,
    self.context.config,
);

gen.convert_to_dart(
    field.try_name_mirror(field.name.rust_style().to_owned()),
)
```
we would have to rewrite the logic both in `ty_enum` and `ty_struct`, and they right now use `field.name` everywhere. Then essentially type needs to know its field or the field function passes itself along to the type? 

IMHO this backlink sounds rather ugly, but maybe I’m missing sth in the code

Comment (fzyzcjy) [reply frb_codegen/src/generator/rust/ty_enum.rs]: I see the problem: The bug is caused by missing a newtype wrapper, so the current way may be OK, let me check it.

File: frb_codegen/src/ir/field.rs [hunk 4]
```diff
@@ -64,4 +72,15 @@ impl IrField {
 
         format!("{enum_name}.{variant_name}")
     }
+
+    /// Takes the provided name and surrounds it with the `mirror_` prefix if the field is in a
+    /// mirrored enum.
+    pub fn try_name_mirror(&self, name: String) -> String {
+        let mirrored_name = self.ty.mirrored_nested();
+        if self.settings.is_in_mirrored_enum && mirrored_name.is_some() {
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:80]: Wait a minute, maybe we already find the answer: Is it possible we simply remove `is_in_mirrored_enum` flag, because is_in_mirrored_enum can be derived from `mirrored_name.is_some()`?

Comment (alexthe2) [reply frb_codegen/src/ir/field.rs:80]: I hoped so as well, but I did a test and sadly there were edge cases in which it was not. I also added a few more unit tests for this

Comment (alexthe2) [reply frb_codegen/src/ir/field.rs:80]: I will try to fix `Vec<Enums>` at one point, so maybe then I’ll find a better way for it

**Comment by alexthe2** [comment]
> (notes for myself)
> 
> Goal: When generating into-dart logic, if the type of a field is a mirror, we should add `mirror_SomeType(...)` wrapper around it.
> 
> Current code: Add "is-mirror" information into the IrField struct.
> 
> Suggestion: Derive "is-mirror" information from the IrType.

small extension to this: This became a new bug, I don’t exactly know what version caused it, my guess is either 1.70, 1.71 or 1.72, because it worked before

**Comment by fzyzcjy** [comment]
I see... Then maybe hack like this temporarily :(

**Comment by alexthe2** [comment]
> I see... Then maybe hack like this temporarily :( Since I do not have more time for this edge case

any nits or can we merge?

File: frb_codegen/src/ir/field.rs [hunk 5]
```diff
@@ -64,4 +72,15 @@ impl IrField {
 
         format!("{enum_name}.{variant_name}")
     }
+
+    /// Takes the provided name and surrounds it with the `mirror_` prefix if the field is in a
+    /// mirrored enum.
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:77]:
```suggestion
    /// Takes the provided name and surrounds it with the `mirror_` prefix if the field is in a
    /// mirrored enum.
    /// This may be improved in the future: https://github.com/fzyzcjy/flutter_rust_bridge/pull/1144#issuecomment-1486569869
```

**Review by fzyzcjy — APPROVED** [review]

**Comment by fzyzcjy** [comment]
No nits, good job!

**Comment by alexthe2** [comment]
Cool, the release will be out once the release pipelines pass, won't it?

**Comment by fzyzcjy** [comment]
Executing release pipeline on my machine :)

image

**Comment by fzyzcjy** [comment]
Ping me if it does not appear after an hour since this one usually takes only a dozen (?) minutes

**Comment by alexthe2** [comment]
Oh ok, thank you

**Comment by fzyzcjy** [comment]
Indeed done now!

You are welcome :)

image
